feat(image): Add image-src flag to specify which runtime(s) to use

pmengelbert opened the pull request:
## Description

Enable a new trivy image `--image-src` flag. For example

**before**:
```bash
trivy image alpine:3.7.3 # tries docker, containerd, podman, remote in that order (hardcoded)
```

**after**:
```bash
trivy image alpine:3.7.3 # behavior is the same as in the *before* example
trivy image --image-src=podman,containerd alpine:3.7.3 # tries only podman, containerd in the specified order
```

## Related issues
- Close #3049 
- #4044
- Close #3929 

## Related PRs
- [ ] Replaces #4015 

Remove this section if you don't have related PRs.

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [ ] I've added tests that prove my fix is effective or that my feature works. (*there were existing tests that hit this codepath, but no new tests written*)
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [x] I've added usage information (if the PR introduces new options)
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change).






=======
## Description

## Related issues


@knqyf263 This is what I got done today on the socket selection. Feel free to use this as a starting point for your future work to include this feature in 0.41.0

Opening as a draft PR since I have not written tests or docs for this yet.

knqyf263 commented:
@DmitriyLewen I think we may want to replace [your PR](https://github.com/aquasecurity/trivy/pull/4015) with this approach. Can you take a look?

Also, I suggested as below.
https://github.com/aquasecurity/trivy/issues/3049#issuecomment-1506349642

DmitriyLewen reviewed (commented):
@pmengelbert Thanks for your work!

Looks good. I added some comments. Can you take a look?

pmengelbert commented on pkg/fanal/image/image.go:
```diff
@@ -44,6 +45,26 @@ func DisableRemote() Option {
+
+	return opts
+}
+
```
Thanks for the suggestions! I implemented them with some slight changes. Please let me know what you think and hit resolve if no further comments

DmitriyLewen replied:
Nice work!

DmitriyLewen commented on docs/docs/references/configuration/cli/trivy_filesystem.md:
```diff
@@ -59,6 +59,7 @@ trivy filesystem [flags] PATH
       --registry-token string       registry token
       --rekor-url string            [EXPERIMENTAL] address of rekor STL server (default "https://rekor.sigstore.dev")
       --reset                       remove all caches and database
+      --runtime strings             Runtime(s) to use (docker,containerd,podman,remote) (default [docker,containerd,podman,remote])
```
we need this option only for `image` mode.
You can disable flag for other modes.
https://github.com/aquasecurity/trivy/blob/cc18f92cf39534a2220883722fb26dc1a9e27fda/pkg/commands/app.go#L299

pmengelbert replied:
I moved this flag over to the `image` group, and regenerated the docs.

DmitriyLewen replied:
It make sense. Good solution.

DmitriyLewen commented on docs/docs/references/configuration/cli/trivy_image.md:
```diff
@@ -81,6 +81,7 @@ trivy image [flags] IMAGE_NAME
       --removed-pkgs                   detect vulnerabilities of removed packages (only for Alpine)
       --report string                  specify a format for the compliance report. (default "summary")
       --reset                          remove all caches and database
+      --runtime strings                Runtime(s) to use (docker,containerd,podman,remote) (default [docker,containerd,podman,remote])
```
```suggestion
      --runtime strings                Runtime(s) to use (docker,containerd,podman,remote) (default priority [docker,containerd,podman,remote])
```

DmitriyLewen replied:
We can set multiple runtimes
```suggestion
      --runtimes strings                Runtime(s) to use (docker,containerd,podman,remote) (default [docker,containerd,podman,remote])
```

pmengelbert replied:
Both of these are taken care of.

The format `(default <...>)` in the CLI help is unalterable. To keep docs consistent with the CLI help, I have the following:
```
Runtime(s) to use, in priority order (docker,containerd,podman,remote) (default [docker,containerd,podman,remote])
```

DmitriyLewen commented on pkg/fanal/image/image.go:
```diff
@@ -9,102 +9,52 @@ import (
-	}
-	for _, opt := range opts {
-		opt(o)
+func NewContainerImage(ctx context.Context, imageName string, opt types.RemoteOptions, tryRuntimes []RuntimeFunc) (types.Image, func(), error) {
```
wdyt?
```suggestion
func NewContainerImage(ctx context.Context, imageName string, opt types.RemoteOptions, runtimes []RuntimeFunc) (types.Image, func(), error) {
```

DmitriyLewen commented on pkg/fanal/test/integration/registry_test.go:
```diff
@@ -1,6 +1,3 @@
-//go:build integration
-// +build integration
-
```
It is integration test. Don't need to remove this.

DmitriyLewen replied:
if I understand correctly - if you return this, then broken test will be fixed.

pmengelbert replied:
I restored this, but the same test is failing when I run `mage test:unit`. Actually, it's failing for me (locally) even on the `main` branch.

DmitriyLewen replied:
It should help with broken test
```suggestion
			runtimes, err := WithRuntimes(types.AllRuntimes)
			assert.NoError(t, err)
			img, cleanup, err := NewContainerImage(context.Background(), tt.args.imageName, tt.args.option, runtimes)
```

pmengelbert replied:
Unit tests are passing now, thanks!

DmitriyLewen reviewed (commented):
LGTM
I added some comments. 

Take a look, please.

Also i think we need add some information about this in docs - https://github.com/aquasecurity/trivy/blob/main/docs/docs/target/container_image.md#supported
i mean default value, priority, etc...

pmengelbert commented:
> LGTM I added some comments.
> 
> Take a look, please.
> 
> Also i think we need add some information about this in docs - https://github.com/aquasecurity/trivy/blob/main/docs/docs/target/container_image.md#supported i mean default value, priority, etc...

I added some docs in `container_image.md:

diff --git a/docs/docs/target/container_image.md b/docs/docs/target/container_image.md
index 7350a6a93..108c667a9 100644
--- a/docs/docs/target/container_image.md
+++ b/docs/docs/target/container_image.md
@@ -224,6 +224,23 @@ GitHub Personal Access Token
     You can see environment variables with `docker inspect`.
 
 ## Supported
+
+Trivy will look for the specified image in a series of locations. By default, it
+will first look in the local Docker Engine, then Containerd, Podman, and
+finally container registry.
+
+This behavior can be modified with the `--runtimes` flag. For example, the
+command
+
+```bash
+trivy image --runtimes podman,containerd alpine:3.7.3
+```
+
+Will first search in Podman. If the image is found there, it will be scanned
+and the results returned. If the image is not found in Podman, then Trivy will
+search in Containerd. If the image is not found there either, the scan will
+fail and no more runtimes will be searched.
+
 ### Docker Engine
 Trivy tries to looks for the specified image in your local Docker Engine.
 It will be skipped if Docker Engine is not running locally.

DmitriyLewen commented:
Good work! Can you resolve conflicts?

pmengelbert commented:
@DmitriyLewen Merge conflicts have been resolved, thanks!

AnaisUrlichs reviewed (approved):
docs looks good to me, thank you

knqyf263 commented:
```--runtimes strings Runtime(s) to use, in priority order (docker,containerd,podman,remote) (default [docker,containerd,podman,remote])
```

`remote` runtime sounds like a remote Docker daemon or other daemons. How about `--image-src`, `--image-from`, or something like that? `docker,containerd,podman,registry`.

knqyf263 commented on pkg/types/target.go:
```diff
@@ -14,6 +14,12 @@ type Scanner string
+type Runtime string
+
+// Runtimes is a slice of runtimes
+type Runtimes []Runtime
```
It should be defined under ImageOptions.
https://github.com/aquasecurity/trivy/blob/9bc326909fce6f5a591eaded17a26a0560d4ad8d/pkg/fanal/types/image.go#L17-L22

pmengelbert replied:
done

knqyf263 commented on pkg/types/target.go:
```diff
@@ -14,6 +14,12 @@ type Scanner string
+type Runtime string
+
+// Runtimes is a slice of runtimes
+type Runtimes []Runtime
```
It should be defined under ImageOptions.
https://github.com/aquasecurity/trivy/blob/9bc326909fce6f5a591eaded17a26a0560d4ad8d/pkg/fanal/types/image.go#L17-L22

pmengelbert replied:
done

knqyf263 commented on pkg/fanal/image/image.go:
```diff
@@ -9,102 +9,56 @@ import (
-		opts.podman = false
-	}
-}
+func WithRuntimes(runtimes ftypes.Runtimes) ([]RuntimeFunc, error) {
```
I think it can be done in NewContainerImage. Is there any benefit in exporting this function?

pmengelbert replied:
I moved this to `NewContainerImage`, and made the function private. The function was also renamed

pmengelbert commented:
> ```
> --runtimes strings Runtime(s) to use, in priority order (docker,containerd,podman,remote) (default [docker,containerd,podman,remote])
> ```
> 
> `remote` runtime sounds like a remote Docker daemon or other daemons. How about `--image-src`, `--image-from`, or something like that? `docker,containerd,podman,registry`.

Renamed to `image-src`, and in the code things like `Runtime` have been renamed to `ImageSource`

pmengelbert commented:
@knqyf263 Is there anything else that this PR needs to be accepted?

knqyf263 commented:
@pmengelbert I'm back from vacation. We plan to take another look at this PR next week. Thanks for your patience.
